tests/test_dependency_manager.py

```diff
@@ -0,0 +1,155 @@
+    dists = attr.ib()  # type: Tuple[FingerprintedDistribution, ...]
+
+    def dist(self, name):
+        # type: (str) -> FingerprintedDistribution
+        project_name = ProjectName(name)
+        dists = [
+            dist for dist in self.dists if project_name == dist.distribution.metadata.project_name
+        ]
+        assert len(dists) == 1, "Expected {name} to match one dist, found {found}".format(
+            name=name, found=" ".join(map(str, dists)) if dists else "none"
+        )
+        return dists[0]
+
+
+@pytest.fixture
+def dist_graph(dist_factory):
+    # type: (DistFactory) -> DistGraph
+
+    # distA     distB <--------\
+    #      \   /     \         |
+    #       v v       v        |
+    #      distC     distD  (cycle)
+    #     /     \   /          |
+    #    V       v v           |
+    # distE     distF ---------/
+
+    return DistGraph(
+        root_reqs=(Requirement.parse("a"), Requirement.parse("b")),
+        dists=(
+            dist_factory.create("A", "c"),
+            dist_factory.create("B", "c", "d"),
+            dist_factory.create("C", "e", "f"),
+            dist_factory.create("D", "f"),
+            dist_factory.create("E"),
+            dist_factory.create("F", "b"),
+        ),
+    )
+
+
+def test_exclude_root_reqs(dist_graph):
+    # type: (DistGraph) -> None
+
+    dependency_manager = DependencyManager(
+        requirements=OrderedSet(dist_graph.root_reqs), distributions=OrderedSet(dist_graph.dists)
+    )
+
+    pex_info = PexInfo.default()
+    pex_builder = PEXBuilder(pex_info=pex_info)
+
+    with warnings.catch_warnings(record=True) as events:
+        dependency_manager.configure(pex_builder, excluded=["a", "b"])
+    assert 2 == len(events)
+
+    warning = events[0]
+    assert PEXWarning == warning.category
+    assert (
+        "The distribution A 0.1.0 was required by the input requirement a but excluded by "
+        "configured excludes: a"
+    ) == str(warning.message)
+
+    warning = events[1]
+    assert PEXWarning == warning.category
+    assert (
+        "The distribution B 0.1.0 was required by the input requirement b but excluded by "
+        "configured excludes: b"
+    ) == str(warning.message)
+
+    pex_builder.freeze()
+
+    assert ["a", "b"] == list(pex_info.requirements)
+    assert ["a", "b"] == list(pex_info.excluded)
+    assert {} == pex_info.distributions
+
+
+def test_exclude_complex(dist_graph):
+    # type: (DistGraph) -> None
+
+    dependency_manager = DependencyManager(
+        requirements=OrderedSet(dist_graph.root_reqs), distributions=OrderedSet(dist_graph.dists)
+    )
+
+    pex_info = PexInfo.default()
+    pex_builder = PEXBuilder(pex_info=pex_info)
+    dependency_manager.configure(pex_builder, excluded=["c"])
+    pex_builder.freeze()
+
+    assert ["a", "b"] == list(pex_info.requirements)
+    assert ["c"] == list(pex_info.excluded)
+    expected_dists = [dist_graph.dist(name) for name in ("A", "B", "D", "F")]
```

Comment:
Double checking my understanding of the behaviour, if `"c"` _wasn't_ excluded, would the expected result be all of the letters, including `C` and `E`?

Reply:
Correct. That's implicitly checked by the 100s of tests that already build and exercise PEXes elsewhere.